Update interpreter semantics of && and || logic operators to be consistent with other logic ops

HeikoBecker opened the pull request:
As described in #651 ,the Nix semantics of `&&` and `||` are inconsistent with those of the other logic operators `==` and `!=`. This PR tries to address this issue by changing their semantics.

The major changes are in ExprSimpleTypesInterpreter ([node url](http://127.0.0.1:63320/node?ref=r%3A78e88ebb-2d27-4b89-867f-623c50619338%28org.iets3.core.expr.simpleTypes.interpreter.plugin%29%2F553080662195536230)).
The PR also adds  tests for the NixSupport of these operators to the nix testcase ([node url](http://127.0.0.1:63320/node?ref=r%3A6731e0cf-cb01-4172-a9af-dc28efa9d184%28test.in.expr.os.nix%40tests%29%2F5987173703616019431)).
These tests mirror the tests described in the `logicaloperators` testcase.

Before the PR, the code for `&&` and `||` did not consistently return `empty` , whereas both `==` and `!=` always evaluate to `empty` if either argument is a NixValue. This PR makes the behavior consistent.

The PR further adds a testsolution `test.nixHandler.expr.os` which ensures that it is also possible to implement short-circuit evaluation with a NixSupport implementation.

Slack discussion for reference: https://itemis.slack.com/archives/CCHR0PVHD/p1676362663689259

arimer commented:
@HeikoBecker the target branch [maintenance/mps20211](https://github.com/IETS3/iets3.opensource/tree/maintenance/mps20211) is no longer under maintenance. Do you mind updating it to `maintenance/mps20213`?

HeikoBecker commented:
> @HeikoBecker the target branch [maintenance/mps20211](https://github.com/IETS3/iets3.opensource/tree/maintenance/mps20211) is no longer under maintenance. Do you mind updating it to `maintenance/mps20213`?
 
Thanks for noticing. I have rebased the branch correctly and updated the target branch.

We can properly review this now.

HeikoBecker commented:
The PR now targets the more recent mps 2022.2 maintenance branch.

arimer reviewed (changes requested):
Just some minor findings we should clarify/address before merging: 

1. TestNixHandlerExtension and TestNixHandler are shown in the diff, although the model seem not to be present. Could you crosscheck if it was removed properly?
2. org.iets3.core.expr.shortcircuit is an empty langauge containing only an extension to the extension point nixHandler. Do we really need a new lanague for it or can we place the code from the plugin model somewhere else? 
3. [NixSupport.get(index)](http://127.0.0.1:63320/node?ref=r%3Aac28053f-2041-47f6-806b-ecfaca05a64a%28org.iets3.core.expr.base.runtime.runtime%29%2F3889855429451896877) and [NixSupport.getReplaced(index)](http://127.0.0.1:63320/node?ref=r%3Aac28053f-2041-47f6-806b-ecfaca05a64a%28org.iets3.core.expr.base.runtime.runtime%29%2F3889855429449865908) are both public and one is calling the other. Do we really need both methods being public?
4. I don't think we need this test here [nixHandler_logand](http://127.0.0.1:63320/node?ref=r%3Ab1d5d1fb-c08e-4ad4-ad15-50e8bc8ef598%28test.nixHandler.expr.os.shortcircuit%40tests%29%2F6506741406319224863)because both cases are already covered here [logicalgroups](http://127.0.0.1:63320/node?ref=r%3Ad9f549b0-d02b-4dca-bb87-3aade9e4e8a5%28test.in.expr.os.operatorgroup%40tests%29%2F8006404979726432233) do we have any difference in the nix declaration?
 <img width="668" alt="grafik" src="https://github.com/IETS3/iets3.opensource/assets/16648247/699182c7-5520-4a5b-905a-1f422f2a8db6">
5.  also the execution of [NixHandlerCanShortCircuit](http://127.0.0.1:63320/node?ref=r%3Ab1d5d1fb-c08e-4ad4-ad15-50e8bc8ef598%28test.nixHandler.expr.os.shortcircuit%40tests%29%2F108752353298805456) does not work for me locally.

arimer commented:
@HeikoBecker I've updated the build script and ran some panding migrations. Hope this is OK for you.

HeikoBecker commented:
> Just some minor findings we should clarify/address before merging:
> 
> 1. TestNixHandlerExtension and TestNixHandler are shown in the diff, although the model seem not to be present. Could you crosscheck if it was removed properly?

I could not find any references locally. They should be gone now.

> 2. org.iets3.core.expr.shortcircuit is an empty langauge containing only an extension to the extension point nixHandler. Do we really need a new lanague for it or can we place the code from the plugin model somewhere else?

Plugin was moved to a model alongside the tests

> 3. [NixSupport.get(index)](http://127.0.0.1:63320/node?ref=r%3Aac28053f-2041-47f6-806b-ecfaca05a64a%28org.iets3.core.expr.base.runtime.runtime%29%2F3889855429451896877) and [NixSupport.getReplaced(index)](http://127.0.0.1:63320/node?ref=r%3Aac28053f-2041-47f6-806b-ecfaca05a64a%28org.iets3.core.expr.base.runtime.runtime%29%2F3889855429449865908) are both public and one is calling the other. Do we really need both methods being public?

I cannot say this for sure since both methods were not introduced by me. I suggest we keep this to a separate PR.

> 4. I don't think we need this test here [nixHandler_logand](http://127.0.0.1:63320/node?ref=r%3Ab1d5d1fb-c08e-4ad4-ad15-50e8bc8ef598%28test.nixHandler.expr.os.shortcircuit%40tests%29%2F6506741406319224863)because both cases are already covered here [logicalgroups](http://127.0.0.1:63320/node?ref=r%3Ad9f549b0-d02b-4dca-bb87-3aade9e4e8a5%28test.in.expr.os.operatorgroup%40tests%29%2F8006404979726432233) do we have any difference in the nix declaration?

Agreed. I removed the test.

> 5. also the execution of [NixHandlerCanShortCircuit](http://127.0.0.1:63320/node?ref=r%3Ab1d5d1fb-c08e-4ad4-ad15-50e8bc8ef598%28test.nixHandler.expr.os.shortcircuit%40tests%29%2F108752353298805456) is using some deprecated API

Fixed in latest commit.

HeikoBecker commented:
I have also rebased and the PR should be ready now.

arimer commented on code/languages/org.iets3.opensource/.mps/libraries.xml:
```diff
@@ -7,15 +7,15 @@
               <option name="name" value="mbeddr.plattform" />
-              <option name='path' value='$PROJECT_DIR$/../../../build/dependencies/com.mbeddr.platform' />            
-   			</Library>
+              <option name="path" value="${iets3.github.opensource.home}/build/dependencies/com.mbeddr.platform" />
```
Which is still fine, because our build.gradle file is already setting it for us...

arimer reviewed (approved):
Let's go 🚀
